# EXP 3 - Mean, Median, Mode (without built-in functions)
**Dataset:** Sub_Division_IMD_2017.csv (IMD subdivision-wise monthly rainfall, India)
**Use case:** KJS-AGR-01 (AI Irrigation Advisory) - same dataset used in EXP1 & EXP2
**Roll No:** 16014324020

In [1]:
import pandas as pd

df = pd.read_csv("Sub_Division_IMD_2017.csv")

# single-attribute sample: ANNUAL rainfall for Konkan & Goa, 1901-2017 (117 values)
konkan = df[df['SUBDIVISION'] == 'Konkan & Goa'].sort_values('YEAR').reset_index(drop=True)
data = konkan['ANNUAL'].tolist()

print("Sample size (n):", len(data))
print("First 10 values:", data[:10])

Sample size (n): 117
First 10 values: [2792.9, 2673.9, 3084.3, 2393.7, 1682.8, 2418.9, 2766.6, 2735.5, 2860.5, 2544.0]


## Mean - sum of all values / number of values

In [2]:
def manual_mean(values):
    total = 0
    for x in values:
        total += x
    return total / len(values)

mean_val = manual_mean(data)
print("Mean =", round(mean_val, 2))

Mean = 2987.53


## Median - middle value after sorting

In [3]:
def manual_median(values):
    sorted_vals = sorted(values)
    n = len(sorted_vals)
    mid = n // 2
    if n % 2 == 0:
        return (sorted_vals[mid - 1] + sorted_vals[mid]) / 2
    else:
        return sorted_vals[mid]

median_val = manual_median(data)
print("Median =", round(median_val, 2))

Median = 2968.5


## Mode - most frequently occurring value

In [4]:
def manual_mode(values):
    counts = {}
    for x in values:
        counts[x] = counts.get(x, 0) + 1

    best_val, best_count = None, 0
    for val, c in counts.items():
        if c > best_count:
            best_val, best_count = val, c
    return best_val, best_count

mode_val, mode_count = manual_mode(data)
print("Mode =", mode_val, " (occurs", mode_count, "time(s))")

Mode = 2792.9  (occurs 1 time(s))


## Final Results

In [5]:
print("Mean   =", round(mean_val, 2))
print("Median =", round(median_val, 2))
print("Mode   =", mode_val, " (occurs", mode_count, "time(s))")

Mean   = 2987.53
Median = 2968.5
Mode   = 2792.9  (occurs 1 time(s))


**Observation:** Mean (2987.53) and Median (2968.5) are close together, so the ANNUAL rainfall for
Konkan & Goa is fairly symmetric with no big outlier years. Mode isn't very meaningful here since rainfall
is continuous data - almost no two years have the exact same value, so every value occurs once.